native_client/c_speech_features/c_speech_features.c

```diff
@@ -0,0 +1,264 @@
+#include <float.h>
+#include "c_speech_features.h"
+#include "tools/kiss_fftr.h"
+
+#define MAX(x,y) ((x) > (y) ? (x) : (y))
+#define MIN(x,y) ((x) < (y) ? (x) : (y))
+
+int
+csf_mfcc(const short* aSignal, unsigned int aSignalLen, int aSampleRate,
+         float aWinLen, float aWinStep, int aNCep, int aNFilters, int aNFFT,
+         int aLowFreq, int aHighFreq, float aPreemph, int aCepLifter,
+         int aAppendEnergy, float*** aMFCC)
+{
+  int i, j, k;
+  float** feat;
+  float* energy;
+
+  int n_frames = csf_fbank(aSignal, aSignalLen, aSampleRate, aWinLen, aWinStep,
+                           aNFilters, aNFFT, aLowFreq, aHighFreq, aPreemph,
+                           &feat, &energy);
+
+  // Perform DCT-II
+  float sf1 = sqrtf(1 / (4 * (float)aNFilters));
+  float sf2 = sqrtf(1 / (2 * (float)aNFilters));
+  float** mfcc = (float**)malloc(sizeof(float*) * n_frames);
+  for (i = 0; i < n_frames; i++) {
+    for (j = 0; j < aNFilters; j++) {
+      feat[i][j] = logf(feat[i][j]);
+    }
+    mfcc[i] = (float*)calloc(sizeof(float), aNCep);
+    for (j = 0; j < aNCep; j++) {
+      for (k = 0; k < aNCep; k++) {
+        mfcc[i][j] += feat[i][k] *
+          cosf(M_PI * j * (2 * k + 1) / (2 * aNFilters));
+      }
+      mfcc[i][j] *= 2 * ((i == 0 && j == 0) ? sf1 : sf2);
+    }
+  }
+
+  // Apply a cepstral lifter
+  csf_lifter(mfcc, n_frames, aNCep, aCepLifter);
+
+  // Append energies
+  if (aAppendEnergy) {
+    for (i = 0; i < n_frames; i++) {
+      mfcc[i][0] = logf(energy[i]);
+    }
+  }
+
+  // Free unused arrays
+  for (i = 0; i < n_frames; i++) {
+    free(feat[i]);
+  }
+  free(feat);
+  free(energy);
+
+  // Return MFCC features
+  *aMFCC = mfcc;
+
+  return n_frames;
+}
+
+int
+csf_fbank(const short* aSignal, unsigned int aSignalLen, int aSampleRate,
+          float aWinLen, float aWinStep, int aNFilters, int aNFFT,
+          int aLowFreq, int aHighFreq, float aPreemph,
+          float*** aFeatures, float** aEnergy)
+{
+  int i, j, k;
+  float** feat;
+  float** fbank;
+  float** pspec;
+  float** frames;
+  float* energy;
+  float* preemph = csf_preemphasis(aSignal, aSignalLen, aPreemph);
+  int frame_len = (int)roundf(aWinLen * aSampleRate);
+  int frame_step = (int)roundf(aWinStep * aSampleRate);
+
+  // Frame the signal into overlapping frames
+  int n_frames = csf_framesig(preemph, aSignalLen, frame_len, aNFFT,
+                              frame_step, &frames);
+
+  // Free preemphasised signal buffer
+  free(preemph);
+
+  // Compute the power spectrum of the frames
+  pspec = csf_powspec((const float**)frames, n_frames, aNFFT);
+
+  // Store the total energy in each frame
+  energy = (float*)calloc(sizeof(float), n_frames);
+  for (i = 0; i < n_frames; i++) {
+    for (j = 0; j < aNFFT / 2 + 1; j++) {
+      energy[i] += pspec[i][j];
+    }
+    if (energy[i] == 0.0f) {
+      energy[i] = FLT_MIN;
+    }
+  }
+
+  // Compute the filter-bank energies
+  fbank = csf_get_filterbanks(aNFilters, aNFFT, aSampleRate,
```

Comment:
And again! I clearly didn't check over this function enough... (I'd like to say some of these leaks weren't there before I separated out the code, but I'm not sure I can :))